domain/consensus/datastructures/headersselectedchainstore/headersselectedchainstore.go

```diff
@@ -0,0 +1,245 @@
+		hscs.stagingRemovedByHash[*blockHash] = struct{}{}
+	}
+
+	currentIndex := uint64(0)
+	highestChainBlockIndex, exists, err := hscs.highestChainBlockIndex(dbContext)
+	if err != nil {
+		return err
+	}
+
+	if exists {
+		currentIndex = highestChainBlockIndex - uint64(len(chainChanges.Removed)) + 1
+	}
+
+	for _, blockHash := range chainChanges.Added {
+		hscs.stagingAddedByIndex[currentIndex] = blockHash
+		hscs.stagingAddedByHash[*blockHash] = currentIndex
+		currentIndex++
+	}
+
+	return nil
+}
+
+func (hscs *headersSelectedChainStore) IsStaged() bool {
+	return len(hscs.stagingAddedByHash) != 0
+}
+
+func (hscs *headersSelectedChainStore) Discard() {
+	hscs.stagingAddedByHash = make(map[externalapi.DomainHash]uint64)
+	hscs.stagingRemovedByHash = make(map[externalapi.DomainHash]struct{})
+	hscs.stagingAddedByIndex = make(map[uint64]*externalapi.DomainHash)
+	hscs.stagingRemovedByIndex = make(map[uint64]struct{})
+}
+
+func (hscs *headersSelectedChainStore) Commit(dbTx model.DBTransaction) error {
+	if !hscs.IsStaged() {
+		return nil
+	}
+
+	for hash := range hscs.stagingRemovedByHash {
+		hashCopy := hash
+		err := dbTx.Delete(hscs.hashAsKey(&hashCopy))
+		if err != nil {
+			return err
+		}
+		hscs.cacheByHash.Remove(&hashCopy)
+	}
+
+	for index := range hscs.stagingRemovedByIndex {
+		err := dbTx.Delete(hscs.indexAsKey(index))
+		if err != nil {
+			return err
+		}
+		hscs.cacheByIndex.Remove(index)
+	}
+
+	highestIndex := uint64(0)
+	for hash, index := range hscs.stagingAddedByHash {
+		hashCopy := hash
+		err := dbTx.Put(hscs.hashAsKey(&hashCopy), hscs.serializeIndex(index))
+		if err != nil {
+			return err
+		}
+
+		err = dbTx.Put(hscs.indexAsKey(index), binaryserialization.SerializeHash(&hashCopy))
+		if err != nil {
+			return err
+		}
+
+		hscs.cacheByHash.Add(&hashCopy, index)
+		hscs.cacheByIndex.Add(index, &hashCopy)
+
+		if index > highestIndex {
+			highestIndex = index
+		}
+	}
+
+	err := dbTx.Put(highestChainBlockIndexKey, hscs.serializeIndex(highestIndex))
+	if err != nil {
+		return err
+	}
+
+	hscs.cacheHighestChainBlockIndex = highestIndex
+
+	hscs.Discard()
+	return nil
+}
+
+// Get gets the acceptanceData associated with the given blockHash
+func (hscs *headersSelectedChainStore) GetIndexByHash(dbContext model.DBReader, blockHash *externalapi.DomainHash) (uint64, bool, error) {
+	if index, ok := hscs.stagingAddedByHash[*blockHash]; ok {
+		return index, true, nil
+	}
+
+	if _, ok := hscs.stagingRemovedByHash[*blockHash]; ok {
+		return 0, false, nil
+	}
+
+	if index, ok := hscs.cacheByHash.Get(blockHash); ok {
+		return index.(uint64), true, nil
+	}
+
+	indexBytes, err := dbContext.Get(hscs.hashAsKey(blockHash))
+	if err != nil {
+		return 0, false, err
+	}
+
+	index := hscs.deserializeIndex(indexBytes)
+	hscs.cacheByHash.Add(blockHash, index)
+	return index, true, nil
+}
+
+func (hscs *headersSelectedChainStore) GetHashByIndex(dbContext model.DBReader, index uint64) (*externalapi.DomainHash, bool, error) {
```

Comment:
This appears to do both `Get` and `Has`. Is there any reason as to why this needs to do both instead of just `Get`?